### ISSUE TITLE
[BUG] model.predict() method for MLP/DLRM/DCN models throw exception

### ISSUE THREAD

**Issue by mamtaed8811** [NVIDIA-Merlin/models#328]
### Bug description

model.predict() method for MLP/DLRM/DCN models throws exception when trying to predict probabilities on the test file.

### Steps/Code to reproduce bug

1.  Download the [sample dataset](https://drive.google.com/drive/u/1/folders/1LjLTVKCESUnE5LQizNVMEpKvfs3A6lI1)
2.  Run the following code below. I am getting valueerror whose screenshots have been attached here. 
3.  Code to Reproduce
     ``` import os
        os.environ["TF_GPU_ALLOCATOR"]="cuda_malloc_async"
        import cudf
        import glob
        import gc
        import pandas as pd
        import tensorflow as tf
        import warnings
        warnings.filterwarnings("ignore")
        
        import nvtabular as nvt
        from nvtabular.ops import *
        from fit_transform import workflow_fit_transform
        
        from merlin.schema.tags import Tags
        from merlin.schema import Schema
        
        import merlin.models.tf as mm
        import merlin.models.tf.dataset as tf_dataloader
        
        from merlin.io.dataset import Dataset
        from merlin.schema.io.tensorflow_metadata import TensorflowMetadata
        from merlin.models.tf.blocks.core.aggregation import CosineSimilarity
        
        train_path = '/workspace/data/train/*.parquet'
        valid_path = '/workspace/data/valid/*.parquet'
        test_path = '/workspace/data/test/*.parquet'
        output_path = '/workspace/data/processed'
        model_path = '/workspace/data/processed/models/'
        
        user_id = ["userid"] >> AddMetadata(tags=[Tags.USER_ID]) >> Categorify(freq_threshold=5)
        item_id = ["contentid"] >> AddMetadata(tags=[Tags.ITEM_ID]) >> Categorify(freq_threshold=5)
        targets = ["relevant"] >> AddMetadata(
        tags=[str(Tags.BINARY_CLASSIFICATION), "target"])
        
        add_feat = ["eventtype", "eventStrength", "BusinessArea", "JobPosition","HrRoleName", "Location"] >> nvt.ops.Categorify()
        
        te_feat = (
            ["userid", "contentid"] + add_feat >>
            TargetEncoding(
                ['relevant'],
                kfold=1,
                p_smooth=20
            ) >>
            Normalize()
        )
        
        outputs = user_id+item_id+targets+add_feat+te_feat
        
        workflow = nvt.Workflow(outputs)
        
        train_dataset = nvt.Dataset(train_path)
        valid_dataset = nvt.Dataset(valid_path)
        test_dataset = nvt.Dataset(test_path)
        
        workflow.fit(train_dataset)
        workflow.transform(train_dataset).to_parquet(output_path=output_path + "/train/")
        workflow.transform(valid_dataset).to_parquet(output_path=output_path + "/valid/")
        workflow.transform(test_dataset).to_parquet(output_path=output_path + "/test/",shuffle=False)
        
        workflow.save('workflow')
        
        schema = TensorflowMetadata.from_proto_text_file(output_path + '/train/').to_merlin_schema()
        target_column = schema.select_by_tag(Tags.TARGET).column_names[0]
        
        model2 = mm.MLPBlock([64, 32]).to_model(
            schema, 
            prediction_tasks=mm.BinaryClassificationTask(target_column)
        )
        
        model2_description = "MLP"
        
        
        model3 = mm.DLRMModel(
            schema,
            embedding_dim=64,
            bottom_block=mm.MLPBlock([128, 64]),
            top_block=mm.MLPBlock([128, 64, 32]),
            prediction_tasks=mm.BinaryClassificationTask(target_column)
        )
        
        model3_description = "DLRM"
        
        
        model4 = mm.DCNModel(
            schema,
            depth=2,
            deep_block=mm.MLPBlock([64, 32]),
            prediction_tasks=mm.BinaryClassificationTask(target_column)
        )
        model4_description = "DCN"
        
        model2.compile('SGD', run_eagerly=False)
        model3.compile('SGD', run_eagerly=False)
        model4.compile('SGD', run_eagerly=False)
        
        batch_size = 16*1024
        train_dl = tf_dataloader.BatchedDataset(
            Dataset(output_path + '/train/*.parquet', part_size="500MB"),
            batch_size = batch_size,
            label_names = [target_column],
            shuffle= True,
            schema = schema,
        )
        
        valid_dl = tf_dataloader.BatchedDataset(
            Dataset(output_path + '/valid/*.parquet', part_size="500MB"),
            batch_size = batch_size,
            label_names = [target_column],
            shuffle = False,
            schema = schema,
        )
        model2.fit(train_dl, validation_data=valid_dl)
        model3.fit(train_dl, validation_data=valid_dl)
        model4.fit(train_dl, validation_data=valid_dl)
        
        model2name = model_path + model2_description
        model3name = model_path + model3_description
        model4name = model_path + model4_description
        
        model2.save(model2name)
        model3.save(model3name)
        model4.save(model4name)
        
        test = pd.read_parquet("/workspace/data/processed/test/part_0.parquet")
        test_input = test[['contentid', 'userid', 'eventtype', 'eventStrength','BusinessArea','JobPosition', 'HrRoleName', 'Location']]
        
        def tf_tensor_dict(df):
            data = df.to_dict("list")
            return {key: tf.convert_to_tensor(value) for key, value in data.items()}
        
        test_tensor = tf_tensor_dict(test_input)
        
        test['predict_proba_MLP'] =  @model2.predict(test_tensor)
```

### Environment details
- Merlin version: 
- Platform:
- Python version:
- PyTorch version (GPU?):
- Tensorflow version (GPU?):

### Additional context

![Screen Shot 2022-04-04 at 9 17 34 AM](https://user-images.githubusercontent.com/28987826/161588939-8a61431b-cba3-47e6-926a-0ddbf676caca.png)
![Screen Shot 2022-04-04 at 9 18 00 AM](https://user-images.githubusercontent.com/28987826/161588954-0cf798b4-939e-405e-a31d-aeeca8883592.png)

**Comment by rnyak** [comment]
@mamtaed8811 May be  you can  update your main branch  to explore the latest examples:
```
cd /models
git checkout main
git pull origin main
pip install -e .
```
then you can build your models with the updated nbs (see 03-Exploring-different-models) . 

For the error you are getting, pls get the batch from you dataloader and check if the dtypes of your model input are matching with the dtypes  of the input features in the test_tensor, if not make the dtypes of `test_tensor` as same in your batch.

```
import merlin.models.tf.dataset as tf_dataloader
train_dl= tf_dataloader.BatchedDataset(
        Dataset(output_path + '/train/*.parquet'),   # if you are using new examples this line should be only  `train`.
        batch_size=32,
        shuffle=False,
    )
batch = next(iter(train_dl))
batch
```

**Comment by mamtaed8811** [comment]
@rnyak I don't quite follow your comment inside the code block. 
My follow up questions:
1) ```import merlin.models.tf.dataset as tf_dataloader
train_dl= tf_dataloader.BatchedDataset(
        Dataset(output_path + '/train/*.parquet'),   # if you are using new examples this line should be only  `train`.
        batch_size=32,
        shuffle=False,
    )
batch = next(iter(train_dl))
batch```

In the above code, for my test dataset, am I to reuse it like this below?
``` 
test_dl= tf_dataloader.BatchedDataset(
        Dataset(output_path + '/test/*.parquet'),   # if you are using new examples this line should be only  `train`.
        batch_size=32,
        shuffle=False,
    )
batch = next(iter(test_dl))
batch
```

"# if you are using new examples this line should be only  `train`"

What am I supposed to be passing inside `Dataset` for new dataset i.e test files?

**Comment by mamtaed8811** [comment]
@rnyak I am creating a dataloader for test files like in my screenshot and then reading the actual test parquet file where I am filtering out the unwanted columns and then comparing the dtypes. Obviously, the dtypes won't match because NVTabular is converting the categorical columns into integer type for the batch. My question here is how do I use a stored test parquet file(unseen data for the model) for inference. What's the right way of doing it? This definitely is not the right way as it's not working currently. Could you please reproduce it on your end and guide me on the right way to use the API for inferencing?
![Screen Shot 2022-04-05 at 11 59 37 AM](https://user-images.githubusercontent.com/28987826/161830381-320b39d8-840f-4ee1-b520-7c44542b748f.png)

**Comment by rnyak** [comment]
@mamtaed8811 you need to use the preprocessed test set to feed to model. you cannot feed raw test set to model.  use the one processed from NVTabular.

**Comment by mamtaed8811** [comment]
I am not seeing this issue anymore with the newest container. Thank you! I will close it now.
